File: android/src/main/kotlin/dev/yusufdemir/meta_flutter_sdk/MetaFlutterSdkPlugin.kt
package dev.yusufdemir.meta_flutter_sdk

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import com.facebook.AccessToken
import com.facebook.CallbackManager
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.FacebookSdk
import com.facebook.GraphRequest
import com.facebook.GraphResponse
import com.facebook.HttpMethod
import com.facebook.appevents.AppEventsLogger
import com.facebook.login.LoginManager
import com.facebook.login.LoginResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry

class MetaFlutterSdkPlugin :
    FlutterPlugin,
    MethodChannel.MethodCallHandler,
    ActivityAware,
    PluginRegistry.ActivityResultListener {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var activity: Activity? = null
    private val callbackManager = CallbackManager.Factory.create()
    private var pendingLogin: MethodChannel.Result? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "dev.yusufdemir/meta_flutter_sdk")
        channel.setMethodCallHandler(this)
        registerLoginCallback()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "initialize" -> initialize(call, result)
                "setAutoLogAppEventsEnabled" -> complete(result) {
                    FacebookSdk.setAutoLogAppEventsEnabled(call.required("enabled"))
                }
                "setAdvertiserIdCollectionEnabled" -> complete(result) {
                    FacebookSdk.setAdvertiserIDCollectionEnabled(call.required("enabled"))
                }
                "setAdvertiserTrackingEnabled" -> result.error(
                    "unsupported",
                    "Advertiser tracking is an iOS-only setting. Use advertiser ID collection on Android.",
                    null,
                )
                "setDataProcessingOptions" -> complete(result) {
                    FacebookSdk.setDataProcessingOptions(
                        call.required<List<String>>("options").toTypedArray(),
                        call.required("country"),
                        call.required("state"),
                    )
                }
                "setFlushBehavior" -> setFlushBehavior(call, result)
                "logEvent" -> logEvent(call, result)
                "flushEvents" -> complete(result) { AppEventsLogger.newLogger(context).flush() }
                "login" -> login(call, result)
                "logout" -> complete(result) { LoginManager.getInstance().logOut() }
                "currentAccessToken" -> result.success(accessTokenMap(AccessToken.getCurrentAccessToken()))
                "getAnonymousId" -> result.success(AppEventsLogger.getAnonymousAppDeviceGUID(context))
                "setUserId" -> complete(result) { AppEventsLogger.setUserID(call.argument<String>("userId")) }
                "graphRequest" -> graphRequest(call, result)
                else -> result.notImplemented()
            }
        } catch (error: Throwable) {
            result.nativeError(error)
        }
    }

    private fun initialize(call: MethodCall, result: MethodChannel.Result) {
        val autoInitialize = call.argument<Boolean>("autoInitialize") ?: true
        FacebookSdk.setAutoInitEnabled(autoInitialize)
        call.argument<Boolean>("autoLogAppEventsEnabled")?.let {
            FacebookSdk.setAutoLogAppEventsEnabled(it)
        }
        call.argument<Boolean>("advertiserIdCollectionEnabled")?.let {
            FacebookSdk.setAdvertiserIDCollectionEnabled(it)
        }
        call.argument<String>("flushBehavior")?.let { applyFlushBehavior(it) }

        if (autoInitialize && !FacebookSdk.isFullyInitialized()) {
            FacebookSdk.fullyInitialize()
        }

        result.success(
            mapOf(
                "platform" to "android",
                "sdkVersion" to FacebookSdk.getSdkVersion(),
                "initialized" to FacebookSdk.isFullyInitialized(),
            ),
        )
    }

    private fun setFlushBehavior(call: MethodCall, result: MethodChannel.Result) = complete(result) {
        applyFlushBehavior(call.required("behavior"))
    }

    private fun applyFlushBehavior(behavior: String) {
        AppEventsLogger.setFlushBehavior(
            when (behavior) {
                "automatic" -> AppEventsLogger.FlushBehavior.AUTO
                "explicitOnly" -> AppEventsLogger.FlushBehavior.EXPLICIT_ONLY
                else -> throw IllegalArgumentException("Unknown flush behavior: $behavior")
            },
        )
    }

    private fun logEvent(call: MethodCall, result: MethodChannel.Result) = complete(result) {
        requireInitialized()
        val logger = AppEventsLogger.newLogger(context)
        val name = call.required<String>("name")
        val parameters = bundle(call.argument<Map<String, Any?>>("parameters").orEmpty())
        val valueToSum = call.argument<Double>("valueToSum")
        if (valueToSum == null) {
            logger.logEvent(name, parameters)
        } else {
            logger.logEvent(name, valueToSum, parameters)
        }
    }

    private fun login(call: MethodCall, result: MethodChannel.Result) {
        requireInitialized()
        val currentActivity = activity ?: run {
            result.error("no_activity", "Facebook Login requires a foreground Activity.", null)
            return
        }
        if (pendingLogin != null) {
            result.error("login_in_progress", "Another Facebook Login request is active.", null)
            return
        }
        if ((call.argument<String>("tracking") ?: "enabled") == "limited") {
            result.error("unsupported", "Limited Login is only supported on iOS.", null)
            return
        }
        pendingLogin = result
        LoginManager.getInstance().logInWithReadPermissions(
            currentActivity,
            call.argument<List<String>>("permissions").orEmpty(),
        )
    }

    private fun graphRequest(call: MethodCall, result: MethodChannel.Result) {
        requireInitialized()
        val parameters = bundle(call.argument<Map<String, Any?>>("parameters").orEmpty())
        call.argument<String>("accessToken")?.let { parameters.putString("access_token", it) }
        val method = when (call.required<String>("method")) {
            "GET" -> HttpMethod.GET
            "POST" -> HttpMethod.POST
            "DELETE" -> HttpMethod.DELETE
            else -> throw IllegalArgumentException("Unsupported Graph method.")
        }
        val request = GraphRequest(
            AccessToken.getCurrentAccessToken(),
            call.required("path"),
            parameters,
            method,
        )
        request.callback = GraphRequest.Callback { response -> handleGraphResponse(response, result) }
        request.executeAsync()
    }

    private fun handleGraphResponse(response: GraphResponse, result: MethodChannel.Result) {
        val error = response.error
        if (error != null) {
            result.error(
                "graph_error",
                error.errorMessage ?: "Graph API request failed.",
                mapOf(
                    "type" to error.errorType,
                    "code" to error.errorCode,
                    "subcode" to error.subErrorCode,
                    "body" to response.rawResponse,
                ),
            )
            return
        }
        result.success(
            mapOf(
                "statusCode" to (response.connection?.responseCode ?: 200),
                "body" to (response.rawResponse ?: response.jsonObject?.toString() ?: "null"),
            ),
        )
    }

    private fun registerLoginCallback() {
        LoginManager.getInstance().registerCallback(
            callbackManager,
            object : FacebookCallback<LoginResult> {
                override fun onSuccess(result: LoginResult) {
                    pendingLogin?.success(
                        mapOf(
                            "cancelled" to false,
                            "accessToken" to accessTokenMap(result.accessToken),
                            "authenticationToken" to null,
                            "grantedPermissions" to result.recentlyGrantedPermissions.toList(),
                            "declinedPermissions" to result.recentlyDeniedPermissions.toList(),
                            "error" to null,
                        ),
                    )
                    pendingLogin = null
                }

                override fun onCancel() {
                    pendingLogin?.success(
                        mapOf(
                            "cancelled" to true,
                            "accessToken" to null,
                            "authenticationToken" to null,
                            "grantedPermissions" to emptyList<String>(),
                            "declinedPermissions" to emptyList<String>(),
                            "error" to mapOf(
                                "code" to "login_cancelled",
                                "message" to "The user cancelled Facebook Login.",
                                "details" to null,
                            ),
                        ),
                    )
                    pendingLogin = null
                }

                override fun onError(error: FacebookException) {
                    pendingLogin?.success(
                        mapOf(
                            "cancelled" to false,
                            "accessToken" to null,
                            "authenticationToken" to null,
                            "grantedPermissions" to emptyList<String>(),
                            "declinedPermissions" to emptyList<String>(),
                            "error" to mapOf(
                                "code" to "login_failed",
                                "message" to (error.message ?: "Facebook Login failed."),
                                "details" to mapOf(
                                    "type" to error.javaClass.name,
                                    "cause" to error.cause?.javaClass?.name,
                                ),
                            ),
                        ),
                    )
                    pendingLogin = null
                }
            },
        )
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean =
        callbackManager.onActivityResult(requestCode, resultCode, data)

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        LoginManager.getInstance().unregisterCallback(callbackManager)
    }

    private fun requireInitialized() {
        check(FacebookSdk.isFullyInitialized()) { "Meta SDK is not initialized. Call initialize() first." }
    }

    private fun bundle(values: Map<String, Any?>): Bundle = Bundle().apply {
        values.forEach { (key, value) ->
            when (value) {
                null -> Unit
                is String -> putString(key, value)
                is Boolean -> putBoolean(key, value)
                is Int -> putInt(key, value)
                is Long -> putLong(key, value)
                is Float -> putFloat(key, value)
                is Double -> putDouble(key, value)
                else -> throw IllegalArgumentException("Unsupported parameter type for '$key'.")
            }
        }
    }

    private fun accessTokenMap(token: AccessToken?): Map<String, Any?>? = token?.let {
        mapOf(
            "token" to it.token,
            "userId" to it.userId,
            "applicationId" to it.applicationId,
            "permissions" to it.permissions.toList(),
            "declinedPermissions" to it.declinedPermissions.toList(),
            "expiredPermissions" to it.expiredPermissions.toList(),
            "expiresAt" to it.expires.time,
            "dataAccessExpiresAt" to it.dataAccessExpirationTime.time,
        )
    }

    private fun complete(result: MethodChannel.Result, action: () -> Unit) {
        action()
        result.success(null)
    }

    private inline fun <reified T> MethodCall.required(name: String): T =
        argument<T>(name) ?: throw IllegalArgumentException("Missing argument: $name")

    private fun MethodChannel.Result.nativeError(error: Throwable, code: String = "native_error") {
        error(code, error.message ?: error.javaClass.simpleName, mapOf("type" to error.javaClass.name))
    }
}

File: ios/meta_flutter_sdk/Sources/meta_flutter_sdk/MetaFlutterSdkPlugin.swift
import FBSDKCoreKit
import FBSDKLoginKit
import Flutter
import UIKit

public final class MetaFlutterSdkPlugin: NSObject, FlutterPlugin {
  private let loginManager = LoginManager()
  private weak var viewController: UIViewController?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "dev.yusufdemir/meta_flutter_sdk",
      binaryMessenger: registrar.messenger()
    )
    let instance = MetaFlutterSdkPlugin()
    instance.viewController = registrar.viewController
    registrar.addMethodCallDelegate(instance, channel: channel)
    registrar.addApplicationDelegate(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    do {
      switch call.method {
      case "initialize":
        try initialize(arguments(call), result: result)
      case "setAutoLogAppEventsEnabled":
        Settings.shared.isAutoLogAppEventsEnabled = try required(call, "enabled")
        result(nil)
      case "setAdvertiserIdCollectionEnabled":
        Settings.shared.isAdvertiserIDCollectionEnabled = try required(call, "enabled")
        result(nil)
      case "setAdvertiserTrackingEnabled":
        applyAdvertiserTrackingEnabled(try required(call, "enabled"))
        result(nil)
      case "setDataProcessingOptions":
        Settings.shared.setDataProcessingOptions(
          try required(call, "options"),
          country: Int32(try required(call, "country") as Int),
          state: Int32(try required(call, "state") as Int)
        )
        result(nil)
      case "setFlushBehavior":
        try applyFlushBehavior(try required(call, "behavior"))
        result(nil)
      case "logEvent":
        try logEvent(call)
        result(nil)
      case "flushEvents":
        AppEvents.shared.flush()
        result(nil)
      case "login":
        try login(call, result: result)
      case "logout":
        loginManager.logOut()
        result(nil)
      case "currentAccessToken":
        result(accessTokenMap(AccessToken.current))
      case "getAnonymousId":
        result(AppEvents.shared.anonymousID)
      case "setUserId":
        AppEvents.shared.userID = arguments(call)["userId"] as? String
        result(nil)
      case "graphRequest":
        try graphRequest(call, result: result)
      default:
        result(FlutterMethodNotImplemented)
      }
    } catch {
      result(flutterError(error))
    }
  }

  private func initialize(_ values: [String: Any], result: FlutterResult) throws {
    if let enabled = values["autoLogAppEventsEnabled"] as? Bool {
      Settings.shared.isAutoLogAppEventsEnabled = enabled
    }
    if let enabled = values["advertiserIdCollectionEnabled"] as? Bool {
      Settings.shared.isAdvertiserIDCollectionEnabled = enabled
    }
    if let enabled = values["advertiserTrackingEnabled"] as? Bool {
      applyAdvertiserTrackingEnabled(enabled)
    }
    if let behavior = values["flushBehavior"] as? String {
      try applyFlushBehavior(behavior)
    }

    let shouldInitialize = values["autoInitialize"] as? Bool ?? true
    if shouldInitialize {
      ApplicationDelegate.shared.initializeSDK()
    }
    result([
      "platform": "ios",
      "sdkVersion": Settings.shared.sdkVersion,
      "initialized": shouldInitialize,
    ])
  }

  private func applyFlushBehavior(_ behavior: String) throws {
    switch behavior {
    case "automatic": AppEvents.shared.flushBehavior = .auto
    case "explicitOnly": AppEvents.shared.flushBehavior = .explicitOnly
    default: throw MetaNativeError.invalidArgument("Unknown flush behavior: \(behavior)")
    }
  }

  private func applyAdvertiserTrackingEnabled(_ enabled: Bool) {
    Settings.shared.advertisingTrackingStatus = enabled ? .allowed : .disallowed
  }

  private func logEvent(_ call: FlutterMethodCall) throws {
    let name: String = try required(call, "name")
    let rawParameters = (arguments(call)["parameters"] as? [String: Any]) ?? [:]
    var parameters: [AppEvents.ParameterName: Any] = [:]
    rawParameters.forEach { parameters[AppEvents.ParameterName($0.key)] = $0.value }

    if let value = arguments(call)["valueToSum"] as? Double {
      AppEvents.shared.logEvent(
        AppEvents.Name(name),
        valueToSum: value,
        parameters: parameters
      )
    } else {
      AppEvents.shared.logEvent(AppEvents.Name(name), parameters: parameters)
    }
  }

  private func login(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
    let permissions: [String] = try required(call, "permissions")
    let trackingValue: String = try required(call, "tracking")
    let tracking: LoginTracking = trackingValue == "limited" ? .limited : .enabled
    let nonce = arguments(call)["nonce"] as? String
    let configuration: LoginConfiguration?
    if let nonce = nonce {
      configuration = LoginConfiguration(
        permissions: permissions,
        tracking: tracking,
        nonce: nonce
      )
    } else {
      configuration = LoginConfiguration(permissions: permissions, tracking: tracking)
    }
    guard let configuration = configuration else {
      throw MetaNativeError.invalidArgument("Invalid permissions or nonce.")
    }

    loginManager.logIn(viewController: viewController, configuration: configuration) { loginResult in
      switch loginResult {
      case .cancelled:
        result([
          "cancelled": true,
          "accessToken": NSNull(),
          "authenticationToken": NSNull(),
          "grantedPermissions": [],
          "declinedPermissions": [],
        ])
      case .failed(let error):
        result(self.flutterError(error, code: "login_failed"))
      case .success(let granted, let declined, let token):
        let payload: [String: Any] = [
          "cancelled": false,
          "accessToken": self.accessTokenMap(token) ?? NSNull(),
          "authenticationToken": AuthenticationToken.current?.tokenString ?? NSNull(),
          "grantedPermissions": granted.map(\.name),
          "declinedPermissions": declined.map(\.name),
        ]
        result(payload)
      }
    }
  }

  private func graphRequest(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
    let values = arguments(call)
    let path: String = try required(call, "path")
    let methodValue: String = try required(call, "method")
    let method: HTTPMethod
    switch methodValue {
    case "GET": method = .get
    case "POST": method = .post
    case "DELETE": method = .delete
    default: throw MetaNativeError.invalidArgument("Unsupported Graph method.")
    }
    let request = GraphRequest(
      graphPath: path,
      parameters: values["parameters"] as? [String: Any] ?? [:],
      tokenString: values["accessToken"] as? String ?? AccessToken.current?.tokenString,
      version: nil,
      httpMethod: method
    )
    request.start { _, response, error in
      if let error = error {
        result(self.flutterError(error, code: "graph_error"))
        return
      }
      do {
        let data = try JSONSerialization.data(withJSONObject: response ?? NSNull())
        result([
          "statusCode": 200,
          "body": String(data: data, encoding: .utf8) ?? "null",
        ])
      } catch {
        result(self.flutterError(error, code: "serialization_error"))
      }
    }
  }

  private func accessTokenMap(_ token: AccessToken?) -> [String: Any?]? {
    guard let token = token else { return nil }
    return [
      "token": token.tokenString,
      "userId": token.userID,
      "applicationId": token.appID,
      "permissions": token.permissions.map(\.name),
      "declinedPermissions": token.declinedPermissions.map(\.name),
      "expiredPermissions": token.expiredPermissions.map(\.name),
      "expiresAt": Int(token.expirationDate.timeIntervalSince1970 * 1000),
      "dataAccessExpiresAt": Int(token.dataAccessExpirationDate.timeIntervalSince1970 * 1000),
    ]
  }

  private func arguments(_ call: FlutterMethodCall) -> [String: Any] {
    (call.arguments as? [String: Any]) ?? [:]
  }

  private func required<T>(_ call: FlutterMethodCall, _ key: String) throws -> T {
    guard let value = arguments(call)[key] as? T else {
      throw MetaNativeError.invalidArgument("Missing or invalid argument: \(key)")
    }
    return value
  }

  private func flutterError(_ error: Error, code: String = "native_error") -> FlutterError {
    FlutterError(
      code: code,
      message: error.localizedDescription,
      details: ["type": String(describing: type(of: error))]
    )
  }

  public func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [AnyHashable: Any] = [:]
  ) -> Bool {
    var typedOptions: [UIApplication.LaunchOptionsKey: Any] = [:]
    for (key, value) in launchOptions {
      if let typedKey = key as? UIApplication.LaunchOptionsKey {
        typedOptions[typedKey] = value
      }
    }
    return ApplicationDelegate.shared.application(
      application,
      didFinishLaunchingWithOptions: typedOptions
    )
  }

  public func application(
    _ application: UIApplication,
    open url: URL,
    options: [UIApplication.OpenURLOptionsKey: Any] = [:]
  ) -> Bool {
    ApplicationDelegate.shared.application(application, open: url, options: options)
  }

  @available(iOS 13.0, *)
  public func scene(_ scene: UIScene, openURLContexts URLContexts: Set<UIOpenURLContext>) {
    guard let context = URLContexts.first else { return }
    var options: [UIApplication.OpenURLOptionsKey: Any] = [:]
    if let sourceApplication = context.options.sourceApplication {
      options[.sourceApplication] = sourceApplication
    }
    if let annotation = context.options.annotation {
      options[.annotation] = annotation
    }
    ApplicationDelegate.shared.application(
      UIApplication.shared,
      open: context.url,
      options: options
    )
  }
}

private enum MetaNativeError: LocalizedError {
  case invalidArgument(String)

  var errorDescription: String? {
    switch self {
    case .invalidArgument(let message): return message
    }
  }
}
